libs/wingc/src/dtsify/extern_dtsify.rs

```diff
@@ -0,0 +1,351 @@
+use std::collections::HashMap;
+
+use camino::Utf8PathBuf;
+use const_format::formatcp;
+use itertools::Itertools;
+
+use crate::{
+	ast::{AccessModifier, Phase},
+	dtsify::{ignore_member_phase, TYPE_INFLIGHT_POSTFIX},
+	files::{update_file, FilesError},
+	jsify::codemaker::CodeMaker,
+	type_check::*,
+	SymbolEnv, WINGSDK_ASSEMBLY_NAME, WINGSDK_DURATION,
+};
+
+const DURATION_FQN: &str = formatcp!("{WINGSDK_ASSEMBLY_NAME}.{WINGSDK_DURATION}");
+
+/// Generates a self-contained .d.ts file for a given extern file.
+pub struct ExternDTSifier<'a> {
+	libraries: &'a SymbolEnv,
+	extern_file: &'a Utf8PathBuf,
+	extern_file_env: &'a SymbolEnvOrNamespace,
+	/// The type information for any named types seen along the way that need to be hoisted to the top of the file
+	hoisted_types: CodeMaker,
+	/// Named types ecountered so far that have been hoisted.
+	/// The key is a psuedo-FQN and the value is the actual name to use in the .d.ts file. This name is based on name_counter since there may be duplicates.
+	known_types: HashMap<String, String>,
+	/// Because all types will be in the same namespace, if you types have the same name this map will be used to disambiguate them by incrementing the counter.
+	name_counter: HashMap<String, usize>,
+}
+
+/// Checks if given file has a valid extension to be considered an extern file
+pub fn is_extern_file(file: &Utf8PathBuf) -> bool {
+	if let Some(ext) = file.extension() {
+		match ext {
+			"js" | "cjs" | "mjs" | "jsx" | "ts" | "cts" | "mts" | "tsx" => true,
```

Comment:
Who woulda thunk a single language would result in so many file extensions / formats